numcodecs/compat.py

```diff
@@ -27,80 +25,150 @@
     from functools import reduce
 
 
-def buffer_tobytes(v):
-    """Obtain a sequence of bytes for the memory buffer used by `v`."""
-    if isinstance(v, binary_type):
-        return v
-    elif isinstance(v, np.ndarray):
-        return v.tobytes(order='A')
-    elif PY2 and isinstance(v, array.array):  # pragma: py3 no cover
-        return v.tostring()
-    else:
-        return memoryview(v).tobytes()
-
-
-def buffer_copy(buf, out=None):
+def memory_copy(buf, out=None):
     """Copy the contents of the memory buffer from `buf` to `out`."""
 
     if out is None:
         # no-op
         return buf
 
-    # handle ndarray destination
-    if isinstance(out, np.ndarray):
+    # obtain ndarrays, casting to the same data type
+    buf = ensure_ndarray_from_memory(buf).view('u1')
+    out = ensure_ndarray_from_memory(out).view('u1')
```

Comment:
Why cast to `uint8`?

Reply:
I think it's needed to ensure source and destination line up? Could be wrong though.

Reply:
Sure, casting may make sense. What about casting `buf` to match `out` though?